File: poc/freudian-neopsyke/src/main/kotlin/freudian/poc/instrumentation/EventLogger.kt
package freudian.poc.instrumentation

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import java.io.Closeable
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption
import java.time.Instant

data class RuntimeEvent(
    val timestamp: String = Instant.now().toString(),
    val tick: Int? = null,
    val type: String,
    val attributes: Map<String, Any?> = emptyMap(),
)

interface EventLogger {
    fun log(event: RuntimeEvent)
}

class JsonlEventLogger(
    private val outputPath: Path,
    private val objectMapper: ObjectMapper = jacksonObjectMapper(),
) : EventLogger, Closeable {

    init {
        outputPath.parent?.let { Files.createDirectories(it) }
        Files.deleteIfExists(outputPath)
        Files.createFile(outputPath)
    }

    override fun log(event: RuntimeEvent) {
        val line = objectMapper.writeValueAsString(event)
        Files.writeString(
            outputPath,
            "$line\n",
            StandardOpenOption.APPEND,
        )
    }

    override fun close() {
        // no resources held open
    }
}

class CompositeEventLogger(
    private val delegates: List<EventLogger>,
) : EventLogger {
    override fun log(event: RuntimeEvent) {
        delegates.forEach { it.log(event) }
    }
}

class InMemoryEventLogger : EventLogger {
    private val mutableEvents = mutableListOf<RuntimeEvent>()
    val events: List<RuntimeEvent> get() = mutableEvents

    override fun log(event: RuntimeEvent) {
        mutableEvents += event
    }
}

File: poc/freudian-neopsyke/py/event_logger.py
import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Protocol


@dataclass
class RuntimeEvent:
    type: str
    attributes: dict[str, Any]
    tick: int | None = None

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {"type": self.type}
        if self.tick is not None:
            d["tick"] = self.tick
        d["attributes"] = self.attributes
        return d


class EventLogger(Protocol):
    def log(self, event: RuntimeEvent) -> None: ...
    def close(self) -> None: ...


class JsonlEventLogger:
    def __init__(self, path: Path) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        self._file = open(path, "w")

    def log(self, event: RuntimeEvent) -> None:
        self._file.write(json.dumps(event.to_dict()) + "\n")

    def close(self) -> None:
        self._file.close()


class CompositeEventLogger:
    def __init__(self, loggers: list[JsonlEventLogger]) -> None:
        self._loggers = loggers

    def log(self, event: RuntimeEvent) -> None:
        for logger in self._loggers:
            logger.log(event)

    def close(self) -> None:
        for logger in self._loggers:
            logger.close()


class InMemoryEventLogger:
    def __init__(self) -> None:
        self.events: list[RuntimeEvent] = []

    def log(self, event: RuntimeEvent) -> None:
        self.events.append(event)

    def close(self) -> None:
        pass
